### PR TITLE
Add the ability to scroll searches with this facade

### PR DESCRIPTION
**Pull request by ianmcorvidae**
This will be useful for some folks using this facade for uses other than our usual ones, particularly the iRODS dashboard. This also includes some decently nice refactoring, so that's cool.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,4 +16,8 @@ endpoints
...
+ * /data/search responds to POST requests. It expects a 'user' query parameter, and within the body of the request either:
+   * a querydsl-compatible JSON query inside a JSON object under the key 'query'
+   * a scroll ID under the key 'scroll\_id'
+
+   It optionally accepts integers in the "from" and "size" keys of the body, which specify the offset and the limit respectively. The defaults are 0 for "from" and 10 for "size". It optionally accepts a "sort" key, which should be an array of objects with "field" and "order" subfields, where "order" is either "ascending" or "descending" and the fields are within the values listed in the documentation endpoint. This endpoint automatically filters results to those the user can see, and adds a "permission" field that summarizes the requesting user's effective permission on each result. It optionally accepts a scroll time under the 'scroll' key; when provided it will return a 'scroll\_id' in the output. When 'scroll\_id' is provided, 'scroll' should always be, for the next slice.
```
Comment (psarando) [inline README.md]: Extra comma at the end here?
> 'scroll' should always be`,` for the next slice

**Review by psarando — APPROVED** [review]
I don't know the `Elasticer` API, but from what I can tell, this LGTM :shipit:

File: data/data.go [hunk 2]
```diff
@@ -116,34 +127,105 @@ func extractInt(v map[string]interface{}, field string, default_val int) int {
...
+	if err != nil {
+		r.writer.WriteHeader(http.StatusInternalServerError)
+		r.logAndOutputErr(err)
+		return nil, err
+	}
+	permFieldScript := `
+		String perm = null;
+		for (up in params._source.userPermissions) {
+			for (user in ` + string(usersJson) + `) {
+				if (up.user == user && perm != 'own' && !(up.permission == 'read' && perm == 'write')) {
```
Comment (ash3rz) [inline data/data.go:158]: Can't really comment on the Go code, but seems ok!  Was only curious about the purpose of this permission check.

Comment (ianmcorvidae) [reply data/data.go:158]: Well, this is the thing that calculates the provided user's permission level on each result. `perm` is the current calculated highest permission level. So, if it's already `own`, nothing can make it higher. If it's at the `write` level, then it's fine to replace it as long as it's either `write` (changing nothing) or `own`, which works out to "not `read`". Thus: for each entry in the document, for each user/group that the user belongs to, if the permission entry in the doc is for that user, set the calculated permission to that permission as long as it's the same or a higher level of permission (i.e., if the calculated one isn't already ownership and it's not downgrading a write to a read).

But it's definitely a bit convoluted, lol.

File: README.md (hunk 1 above)
Comment (ianmcorvidae) [reply README.md]: Heh, I see what I was going for and I'll rewrite it. Bit confusing as it's currently written. What I mean is, whenever `scroll_id` is set, `scroll` should also be set.

File: data/data.go (hunk 2 above)
Comment (ianmcorvidae) [reply data/data.go:158]: I think I'll add a comment explaining the script. That seems like the thing to do here.

**Comment by ianmcorvidae** [comment]
Thanks for the reviews! I'll merge it now, but further reviews are welcome if anyone's still looking.

File: data/data.go (hunk 2 above)
Comment (ash3rz) [reply data/data.go:158]: Got it - thanks for the explanation!

**Pull request merged**